From ae52df5f61518c1d53c4afb763da249ac2e2c937 Mon Sep 17 00:00:00 2001
From: Benjamin Lin <benjamin-jw.lin@mediatek.com>
Date: Thu, 30 Jul 2026 15:50:18 +0800
Subject: [PATCH] mtk: mt76: mt7996: Refactor WMM PBC for special cases

Problem:
Potential WMM performance downgrade.

Root Cause:
There are two special cases that may cause WMM performance downgrade.
(1) If multi-AC traffic occurs across different bands, PBC will not be
enabled. Consequently, PLE pages shared among all bands will not be
reserved for the traffic. E.g. BE traffic on 2.4GHz, while VO traffic
on 5GHz.
(2) If mt7992 MT76 only enables PBC on band that has multi-AC traffic,
while leave the others disabled, PLE pages will not be reserved for
traffic on the others. E.g. VO and VI traffic on 2.4GHz, while BE
traffic on 5GHz.

Solution:
(1) Change multi-AC traffic detection from per-PHY to per-DEV.
(2) Enable/disable PBC simultaneously on all bands for mt7992.

Feature:
WMM
PBC (Packet Budget Control)
AC Tail Drop

Signed-off-by: Benjamin Lin <benjamin-jw.lin@mediatek.com>
---
 mt7996/mcu.c    | 191 +++++++++++++++++-------------------------------
 mt7996/mt7996.h |   4 +-
 2 files changed, 69 insertions(+), 126 deletions(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 6c6cce55f..766e7ec96 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -1229,40 +1229,6 @@ mt7996_mcu_rx_thermal_info_query(struct mt7996_dev *dev, struct sk_buff *skb)
 	phy->throttle_state = phy->throttle_level[n->state];
 }
 
-static void
-mt7996_mcu_update_pbc_pkt_cnt(void *priv, u8 *mac, struct ieee80211_vif *vif)
-{
-	unsigned long link_id, valid_links = vif->valid_links ?: BIT(0);
-	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
-	struct mt7996_mcu_bss_acq_pkt_cnt_event *event = priv;
-	u32 bitmap = le32_to_cpu(event->bss_bitmap);
-
-	rcu_read_lock();
-	for_each_set_bit(link_id, &valid_links, IEEE80211_MLD_MAX_NUM_LINKS) {
-		struct mt7996_vif_link *link;
-		struct mt7996_phy *phy;
-		u8 bss_idx, ac;
-
-		link = (struct mt7996_vif_link *)rcu_dereference(mvif->mt76.link[link_id]);
-		if (!link)
-			continue;
-
-		bss_idx = link->mt76.idx;
-		if (bss_idx >= BSS_ACQ_PKT_CNT_BSS_NUM || !(bitmap & (1 << bss_idx)))
-			continue;
-
-		phy = link->phy;
-		if (!phy)
-			continue;
-
-		for (ac = IEEE80211_AC_VO; ac < IEEE80211_NUM_ACS; ++ac) {
-			u8 lmac_ac = mt76_connac_lmac_mapping(ac);
-			phy->pbc_pkt_cnt[ac] += le32_to_cpu(event->bss[bss_idx].cnt[lmac_ac]);
-		}
-	}
-	rcu_read_unlock();
-}
-
 void mt7996_mcu_wmm_pbc_work(struct work_struct *work)
 {
 #define WMM_PBC_BSS_ALL		0xff
@@ -1331,54 +1297,65 @@ static void mt7996_queue_phy_pbc_work(struct mt7996_phy *phy, bool multi_ac)
 static void
 mt7996_mcu_rx_bss_acq_pkt_cnt(struct mt7996_dev *dev, struct sk_buff *skb)
 {
-	struct mt7996_mcu_bss_acq_pkt_cnt_event *event = (struct mt7996_mcu_bss_acq_pkt_cnt_event *)skb->data;
+	struct mt7996_mcu_bss_acq_pkt_cnt_event *event;
 	struct mt7996_phy *phy;
-	bool multi_ac = false;
-	u8 ac;
+	int i, j, ac_cnt = 0;
+	u32 bitmap;
 
+	event = (struct mt7996_mcu_bss_acq_pkt_cnt_event *)skb->data;
+	bitmap = le32_to_cpu(event->bss_bitmap);
 	/* Update TX packet or PLE page counters */
-	if (event->bss_bitmap == 0) {
+	if (bitmap == 0) {
+		memset(dev->pbc_pg_cnt, 0, sizeof(dev->pbc_pg_cnt));
+
 		mt7996_for_each_phy(dev, phy) {
 			u8 band = phy->mt76->band_idx;
-
-			memset(phy->pbc_pg_cnt, 0, sizeof(phy->pbc_pg_cnt));
-
-			for (ac = IEEE80211_AC_VO; ac < IEEE80211_NUM_ACS; ++ac) {
-				u8 lmac_ac = mt76_connac_lmac_mapping(ac);
-				phy->pbc_pg_cnt[ac] += le32_to_cpu(event->band[band].pg_cnt[lmac_ac]);
+			for (i = IEEE80211_AC_VO; i < IEEE80211_NUM_ACS; i++) {
+				/* PBC only tracks how many ACs have active
+				 * traffic, so specific AC index is irrelevant.
+				 * Hence, omit unnecessary LMAC AC index mapping
+				 * here for simplicity.
+				 */
+				dev->pbc_pg_cnt[i] += le32_to_cpu(event->band[band].pg_cnt[i]);
 			}
 		}
 	} else {
-		mt7996_for_each_phy(dev, phy)
-			memset(phy->pbc_pkt_cnt, 0, sizeof(phy->pbc_pkt_cnt));
+		memset(dev->pbc_pkt_cnt, 0, sizeof(dev->pbc_pkt_cnt));
 
-		ieee80211_iterate_active_interfaces_atomic(dev->mt76.hw,
-				IEEE80211_IFACE_SKIP_SDATA_NOT_IN_DRIVER,
-				mt7996_mcu_update_pbc_pkt_cnt, event);
+		for (i = 0; i < BSS_ACQ_PKT_CNT_BSS_NUM; i++) {
+			if (!(bitmap & BIT(i)))
+				continue;
+
+			for (j = IEEE80211_AC_VO; j < IEEE80211_NUM_ACS; j++) {
+				/* PBC only tracks how many ACs have active
+				 * traffic, so specific AC index is irrelevant.
+				 * Hence, omit unnecessary LMAC AC index mapping
+				 * here for simplicity.
+				 */
+				dev->pbc_pkt_cnt[j] += le32_to_cpu(event->bss[i].cnt[j]);
+			}
+		}
 	}
 
 	/* Check TX packet and PLE page counters to detect multi-AC traffic */
-	mt7996_for_each_phy(dev, phy) {
-		u8 ac_cnt = 0;
-
-		for (ac = IEEE80211_AC_VO; ac < IEEE80211_NUM_ACS; ++ac)
-			if (phy->pbc_pg_cnt[ac] > WMM_PBC_PG_THRESHOLD ||
-			    phy->pbc_pkt_cnt[ac] > WMM_PBC_PKT_THRESHOLD)
-				++ac_cnt;
-
-		if (ac_cnt > 1)
-			multi_ac = true;
+	for (i = IEEE80211_AC_VO; i < IEEE80211_NUM_ACS; i++) {
+		if (dev->pbc_pg_cnt[i] > WMM_PBC_PG_THRESHOLD ||
+		    dev->pbc_pkt_cnt[i] > WMM_PBC_PKT_THRESHOLD)
+			ac_cnt++;
+	}
 
-		/* For MT7992, PBC is per-PHY configured */
-		if (is_mt7992(&dev->mt76))
+	if (is_mt7992(&dev->mt76)) {
+		/* For MT7992, PBC is per-PHY configured, and it should
+		 * be enabled/disabled simultaneously on all PHYs.
+		 */
+		mt7996_for_each_phy(dev, phy)
 			mt7996_queue_phy_pbc_work(phy, ac_cnt > 1);
+	} else {
+		/* For other chips, PBC is per-DEV configured,
+		 * and its state is maintained in first PHY.
+		 */
+		mt7996_queue_phy_pbc_work(&dev->phy, ac_cnt > 1);
 	}
-
-	/* For other chips, PBC is per-DEV configured,
-	 * and its state is maintained in first PHY.
-	 */
-	if (!is_mt7992(&dev->mt76))
-		mt7996_queue_phy_pbc_work(&dev->phy, multi_ac);
 }
 
 int mt7996_mcu_ba_trigger_enable(struct mt7996_dev *dev, u8 enable)
@@ -1657,82 +1634,48 @@ mt7996_mcu_mld_event(struct mt7996_dev *dev, struct sk_buff *skb)
 	}
 }
 
-static void
-mt7996_mcu_uni_update_pbc_pkt_cnt(void *priv, u8 *mac, struct ieee80211_vif *vif)
-{
-	unsigned long link_id, valid_links = vif->valid_links ?: BIT(0);
-	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
-	struct mt7996_mld_sdo_bss_acq_pkt_cnt *event = priv;
-
-	rcu_read_lock();
-	for_each_set_bit(link_id, &valid_links, IEEE80211_MLD_MAX_NUM_LINKS) {
-		struct mt7996_vif_link *link;
-		struct mt7996_phy *phy;
-		u8 bss_idx, ac;
-		u32 bitmap;
-
-		link = (struct mt7996_vif_link *)rcu_dereference(mvif->mt76.link[link_id]);
-		if (!link)
-			continue;
-
-		bss_idx = link->mt76.idx;
-		if (bss_idx >= UNI_CMD_SDO_CFG_BSS_NUM)
-			continue;
-
-		bitmap = le32_to_cpu(event->bitmap[bss_idx / 32]);
-		if (!(bitmap & (1 << (bss_idx % 32))))
-			continue;
-
-		phy = link->phy;
-		if (!phy)
-			continue;
-
-		for (ac = IEEE80211_AC_VO; ac < IEEE80211_NUM_ACS; ++ac) {
-			u8 lmac_ac = mt76_connac_lmac_mapping(ac);
-			phy->pbc_pkt_cnt[ac] += le32_to_cpu(event->pkt_cnt[bss_idx][lmac_ac]);
-		}
-	}
-	rcu_read_unlock();
-}
-
 static void
 mt7996_mcu_uni_bss_acq_pkt_cnt(struct mt7996_dev *dev, struct tlv *tlv)
 {
-	struct mt7996_phy *phy;
-	bool multi_ac = false;
+	struct mt7996_mld_sdo_bss_acq_pkt_cnt *event;
+	int i, j, ac_cnt = 0;
 
-	if (le16_to_cpu(tlv->len) - 4 != sizeof(struct mt7996_mld_sdo_bss_acq_pkt_cnt)) {
+	event = (struct mt7996_mld_sdo_bss_acq_pkt_cnt *)tlv->data;
+	if (le16_to_cpu(tlv->len) - 4 != sizeof(*event)) {
 		dev_err(dev->mt76.dev, "%s: invalid TLV length: %u\n",
 			__func__, le16_to_cpu(tlv->len));
 		return;
 	}
 
 	/* Update TX packet counters */
-	mt7996_for_each_phy(dev, phy)
-		memset(phy->pbc_pkt_cnt, 0, sizeof(phy->pbc_pkt_cnt));
-
-	ieee80211_iterate_active_interfaces_atomic(dev->mt76.hw,
-			IEEE80211_IFACE_SKIP_SDATA_NOT_IN_DRIVER,
-			mt7996_mcu_uni_update_pbc_pkt_cnt, tlv->data);
+	memset(dev->pbc_pkt_cnt, 0, sizeof(dev->pbc_pkt_cnt));
 
-	/* Check TX packet counters to detect multi-AC traffic */
-	mt7996_for_each_phy(dev, phy) {
-		u8 ac, ac_cnt = 0;
+	for (i = 0; i < UNI_CMD_SDO_CFG_BSS_NUM; i++) {
+		u32 bitmap = le32_to_cpu(event->bitmap[i / 32]);
 
-		for (ac = IEEE80211_AC_VO; ac < IEEE80211_NUM_ACS; ++ac)
-			if (phy->pbc_pkt_cnt[ac] > WMM_PBC_PKT_THRESHOLD)
-				++ac_cnt;
+		if (!(bitmap & BIT(i % 32)))
+			continue;
 
-		if (ac_cnt > 1) {
-			multi_ac = true;
-			break;
+		for (j = IEEE80211_AC_VO; j < IEEE80211_NUM_ACS; j++) {
+			/* PBC only tracks how many ACs have active
+			 * traffic, so specific AC index is irrelevant.
+			 * Hence, omit unnecessary LMAC AC index mapping
+			 * here for simplicity.
+			 */
+			dev->pbc_pkt_cnt[j] += le32_to_cpu(event->pkt_cnt[i][j]);
 		}
 	}
 
+	/* Check TX packet counters to detect multi-AC traffic */
+	for (i = IEEE80211_AC_VO; i < IEEE80211_NUM_ACS; i++) {
+		if (dev->pbc_pkt_cnt[i] > WMM_PBC_PKT_THRESHOLD)
+			ac_cnt++;
+	}
+
 	/* Via unified FW command, PBC is per-DEV configured,
 	 * and its state is maintained in first PHY.
 	 */
-	mt7996_queue_phy_pbc_work(&dev->phy, multi_ac);
+	mt7996_queue_phy_pbc_work(&dev->phy, ac_cnt > 1);
 }
 
 static void
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 214edace8..dad9088a9 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -823,8 +823,6 @@ struct mt7996_phy {
 
 	bool wmm_pbc_enable;
 	struct work_struct wmm_pbc_work;
-	u64 pbc_pkt_cnt[IEEE80211_NUM_ACS];
-	u64 pbc_pg_cnt[IEEE80211_NUM_ACS];
 };
 
 struct mt7996_dev {
@@ -972,6 +970,8 @@ struct mt7996_dev {
 
 	struct mt7996_vow_ctrl vow;
 
+	u64 pbc_pkt_cnt[IEEE80211_NUM_ACS];
+	u64 pbc_pg_cnt[IEEE80211_NUM_ACS];
 	u16 pbc_low_bound[WMM_PBC_QUEUE_NUM];
 	u16 pbc_up_bound[WMM_PBC_QUEUE_NUM];
 #ifdef CONFIG_MTK_DEBUG
-- 
2.45.2

